File: src/file_server/vap_parser.js


const fs = require('fs')
const path = require('path')
export const ffprobe = require('ffprobe')
export const ffprobeStatic = require('ffprobe-static')

class Mp4Box {
    constructor(boxType, start, size, content) {
        this.boxType = boxType
        this.start = start
        this.size = size
        this.content = content
    }

}


export function getVapBoxes(file_path) {
    console.log("start parse mp4");
    var fd = fs.openSync(file_path, 'r')
    var allBoxes = new Array()
    var position = 0
    while (true) {
        var boxStart = position;
        var sizeReadBuffer = Buffer.alloc(4)
        var headerSize = 0
        var sizeRead = fs.readSync(fd, sizeReadBuffer, {
            position: position,
            length: 4
        })
        headerSize = headerSize + 4
        if (sizeRead != 4) {
            console.log("read file end or error")
            break;
        }
        position += 4
        var boxSize = sizeReadBuffer.readUInt32BE()
        console.log("boxSize:", boxSize)

        var boxTypeReadBuffer = Buffer.alloc(4)
        var boxTypeRead = fs.readSync(fd, boxTypeReadBuffer, {
            position: position,
            length: 4
        })
        headerSize = headerSize + 4
        if (boxTypeRead != 4) {
            console.log("read file error - 1")
            break;
        }
        position += 4
        var boxType = boxTypeReadBuffer.toString()

        if (boxSize == 1) {
            var largeSizeReadBuffer = Buffer.alloc(8)
            var largeSizeRead = fs.readSync(fd, largeSizeReadBuffer, {
                position: position,
                length: 8
            })
            if (largeSizeRead != 8) {
                console.log("read file error - 2")
                break;
            }
            position += 8
            boxSize = largeSizeReadBuffer.readBigUInt64BE()
            headerSize = headerSize + 8
        }
        var boxContent = ""
        if (boxType == "vapc") {
            console.log('box size is ', boxSize, "headersize is ", headerSize, "position is ", position)
            var vapcLen = boxSize - headerSize
            var contentReadBuffer = Buffer.alloc(vapcLen)
            var contentRead = fs.readSync(fd, contentReadBuffer, {
                position: position,
                length: vapcLen
            })
            console.log("vaplen is ", vapcLen, "content read is ", contentRead);
            position += vapcLen;
            if (contentRead != vapcLen) {
                console.log("read file error - 3")
                break;
            }
            boxContent = contentReadBuffer.toString()
        }
        var box = new Mp4Box(boxType, boxStart, boxSize, boxContent)
        allBoxes.push(box)
        position = boxStart + boxSize
    }
    return allBoxes
}

export function addVapInfoToMp4(filePath, vapJson) {
    console.log("vap json is ", vapJson)
    var allBoxes = getVapBoxes(filePath)
    for (var boxIndex in allBoxes) {
        var box = allBoxes[boxIndex]
        if (box.boxType == "vapc") {
            return -1, "already vap"
        }
    }
    // add a box ftype=vapc data=vapJson to the mp4 file
    var vapJsonStr = JSON.stringify(vapJson)
    var vapJsonBuffer = Buffer.from(vapJsonStr)
    var vapJsonBufferLen = vapJsonBuffer.length
    var vapcBoxSize = 8 + vapJsonBufferLen
    console.log("ready len is ", vapcBoxSize)
    var vapcBoxBuffer = Buffer.alloc(vapcBoxSize)
    vapcBoxBuffer.writeUInt32BE(vapcBoxSize, 0)
    vapcBoxBuffer.write("vapc", 4)
    vapJsonBuffer.copy(vapcBoxBuffer, 8)

    var fileBaseName = path.basename(filePath)
    var fileDir = path.dirname(filePath)
    var tempFilePath = path.join(fileDir, "__temp" + fileBaseName)
    var fd = fs.openSync(tempFilePath, 'w')
    var oldFd = fs.openSync(filePath, 'r')
    var position = 0
    for (var oldBox of allBoxes) {
        var oldBoxBuffer = Buffer.alloc(oldBox.size)
        fs.readSync(oldFd, oldBoxBuffer, {
            position: oldBox.start,
            length: oldBox.size
        })
        fs.writeSync(fd, oldBoxBuffer, {
            position: oldBox.start
        })
        position = oldBox.start + oldBox.size
    }
    fs.writeSync(fd, vapcBoxBuffer, {
        position: position
    });

    let newAllBox = getVapBoxes(tempFilePath);
    var boxOk = false;
    for (var boxIndex in newAllBox) {
        var box = newAllBox[boxIndex];
        if (box.boxType == "vapc") {
            boxOk = true
            break
        }
    }
    fs.closeSync(fd);
    fs.closeSync(oldFd);
    if (boxOk === true) {
        //cover filePath with tempFilePath       
        fs.unlinkSync(filePath)
        fs.renameSync(tempFilePath, filePath)
        return 0, ""
    } else {
        fs.unlinkSync(tempFilePath)
        return -2, "add vapc failed"
    }
    return 0, ""
}

export function getVapInfo(filePath) {
    var allBoxes = getVapBoxes(filePath)
    for (var boxIndex in allBoxes) {
        var box = allBoxes[boxIndex]
        if (box.boxType == "vapc") {
            var vapJson = box.content
            return JSON.parse(vapJson)
        }
    }
    return null

}

export function getFileInfoOfVap(filePath, forceVap = true) {
    return new Promise((resolve, reject) => {
        console.log("get file info of vap: ", filePath)
        var stat = fs.statSync(filePath)
        var file_info = {}
        file_info["size"] = stat.size;
        // is mp4
        if (filePath.endsWith(".mp4")) {
            var vapJson = getVapInfo(filePath);
            if (vapJson != null || forceVap == false) {
                ffprobe(filePath, { path: ffprobeStatic.path })
                    .then(fileMetaData => {
                        if (fileMetaData == undefined || fileMetaData.streams == undefined) {
                            console.log("ffprobe failed: ", filePath)
                            resolve(null);
                        }
                        if (fileMetaData.streams.length > 0) {
                            for (var info in fileMetaData.streams) {
                                var stream = fileMetaData.streams[info]
                                if (stream.codec_type == "video") {
                                    file_info["video_info"] = {
                                        "codec_name": stream.codec_name,
                                        "width": stream.width,
                                        "height": stream.height,
                                        "duration_ts": stream.duration,
                                        "bit_rate": stream.bit_rate,
                                    }
                                    break;
                                }
                            }
                        }
                        file_info["vap_info"] = vapJson
                        file_info["path"] = filePath
                        resolve(file_info);
                    })
                    .catch(error => {
                        console.log("ffprobe failed: ", filePath)
                        resolve(null);
                    });
            } else {
                console.log("ffprobe failed2: ", filePath)
                resolve(null);
            }
        } else {
            console.log("ffprobe failed3: ", filePath)
            resolve(null);
        }
    });
}

File: server/src/vap_parser.c
//
//  vap_parser.c
//  Vap
//
//  Created by 梁金锋 on 2024/4/27.
//

#include "vap_parser.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include "CJSON.h"
#include "string_util.h"
#include <sys/stat.h>
#include "server_util.h"
#include "vap_parser.h"





BoxArray getVapBoxes(char *filePath);

Mp4Box* createMp4Box(char *boxType, uint64_t start, uint64_t size, char *content) {
    Mp4Box *newBox = malloc(sizeof(Mp4Box));
    newBox->boxType = boxType;
    newBox->start = start;
    newBox->size = size;
    newBox->content = content;
    return newBox;
}

// This function is a placeholder. You'll need to implement it according to your needs.
uint8_t* readFromFile(FILE *file, uint64_t position, uint64_t length) {
    // Check if the requested position is beyond the end of the file
    fseek(file, 0, SEEK_END);
    uint64_t fileSize = ftell(file);
    if (position >= fileSize) {
        printf("Position is beyond the end of the file\n");
        return NULL;
    }

    // Check if the requested length goes beyond the end of the file
    if (position + length > fileSize) {
        printf("Requested length goes beyond the end of the file\n");
        return NULL;
    }

    uint8_t *buffer = malloc(length);
    if (buffer == NULL) {
        printf("Memory allocation failed\n");
        return NULL;
    }

    fseek(file, position, SEEK_SET);
    fread(buffer, sizeof(uint8_t), length, file);

    return buffer;
}

char* readCharFromFile(FILE *file, uint64_t position, uint64_t length) {
    fseek(file, 0, SEEK_END);
    uint64_t fileSize = ftell(file);
    if (position >= fileSize) {
        printf("Position is beyond the end of the file\n");
        return NULL;
    }

    // Check if the requested length goes beyond the end of the file
    if (position + length > fileSize) {
        printf("Requested length goes beyond the end of the file\n");
        return NULL;
    }
    char *buffer = malloc(length + 1);
    if (buffer == NULL) {
        printf("Memory allocation failed\n");
        return NULL;
    }

    fseek(file, position, SEEK_SET);
    fread(buffer, sizeof(char), length, file);
    buffer[length] = '\0'; // Null-terminate the string

    return buffer;
}

uint32_t readUInt32BE(uint8_t *buffer) {
    return ((uint32_t)buffer[0] << 24) |
           ((uint32_t)buffer[1] << 16) |
           ((uint32_t)buffer[2] << 8)  |
           (uint32_t)buffer[3];
}

uint64_t readUInt64BE(uint8_t *buffer) {
    return ((uint64_t)buffer[0] << 56) |
           ((uint64_t)buffer[1] << 48) |
           ((uint64_t)buffer[2] << 40) |
           ((uint64_t)buffer[3] << 32) |
           ((uint64_t)buffer[4] << 24) |
           ((uint64_t)buffer[5] << 16) |
           ((uint64_t)buffer[6] << 8)  |
           (uint64_t)buffer[7];
}

char* convertToString(uint8_t *buffer, int length) {
    char *str = malloc(length + 1);
    if (str == NULL) {
        printf("Memory allocation failed\n");
        return NULL;
    }

    memcpy(str, buffer, length);
    str[length] = '\0';  // Null-terminate the string

    return str;
}

BoxArray getVapBoxes(char *file_path) {
    printf("start parse mp4 %s \n", file_path);
    FILE *fd = fopen(file_path, "r");
    Mp4Box **allBoxes = malloc(sizeof(Mp4Box*) * 100); // Assuming max 100 boxes.
    if (allBoxes == NULL) {
        printf("Memory allocation failed\n");
        return (BoxArray){NULL, 0};
    }
    if (fd == NULL) {
        printf("File not found\n");
        return (BoxArray){NULL, 0, 0};
    }
    uint64_t position = 0;
    uint64_t boxCount = 0;
    if (strcmp(file_path, "/Users/liangjinfeng/Downloads/aaaa/__compress_video.mp4") == 0) {
        printf("ready for test");
    }
    while (1) {
        uint64_t boxStart = position;
        uint8_t *sizeReadBuffer = readFromFile(fd, position, 4);
        uint64_t headerSize = 4;
        if (sizeReadBuffer == NULL) {
            printf("read file end or error\n");
            break;
        }
        position += 4;
        uint64_t boxSize = readUInt32BE(sizeReadBuffer);
        free(sizeReadBuffer);
        printf("boxSize: %lld\n", boxSize);
        if (boxSize == 0) {
            printf("read boxsize error - 1\n");
            return (BoxArray){NULL, 0, 1};
            break;
        }

        uint8_t *boxTypeReadBuffer = readFromFile(fd, position, 4);
        headerSize += 4;
        if (boxTypeReadBuffer == NULL) {
            printf("read file error - 1\n");
            break;
        }
        char *boxType = convertToString(boxTypeReadBuffer, 4);
        free(boxTypeReadBuffer);
        position += 4;

        if (boxSize == 1) {
            uint8_t *largeSizeReadBuffer = readFromFile(fd, position, 8);
            if (largeSizeReadBuffer == NULL) {
                printf("read file error - 2\n");
                break;
            }
            position += 8;
            boxSize = readUInt64BE(largeSizeReadBuffer);
            free(largeSizeReadBuffer);
            headerSize += 8;
        }
        char *boxContent = "";
        if (strcmp(boxType, "vapc") == 0) {
            uint64_t vapcLen = boxSize - headerSize;
            char *contentReadBuffer = readCharFromFile(fd, position, vapcLen);
            position += vapcLen;
            if (contentReadBuffer == NULL) {
                printf("read file error - 3\n");
                break;
            }
            boxContent = contentReadBuffer;
        }
        Mp4Box *box = createMp4Box(boxType, boxStart, boxSize, boxContent);
        allBoxes[boxCount++] = box;
        position = boxStart + boxSize;
    }
    fclose(fd);
    BoxArray boxArray = {
        .boxes = allBoxes,
        .length = boxCount
    };
    return boxArray;
}

cJSON *getVapcInfo(char *filePath) {
    cJSON *vapJson = NULL;
    BoxArray array = getVapBoxes(filePath);
    for (int i = 0; i < array.length; i++) {
        Mp4Box *box = array.boxes[i];
        if (strcmp(box->boxType, "vapc") == 0) {
            vapJson = cJSON_Parse(box->content);
            break;
        }
    }
    return vapJson;
}

char * getVapcContent(char *filePath) {
    char *vapJsonContent = NULL;
    BoxArray array = getVapBoxes(filePath);
    for (int i = 0; i < array.length; i++) {
        Mp4Box *box = array.boxes[i];
        if (strcmp(box->boxType, "vapc") == 0) {
            vapJsonContent = box->content;
            break;
        }
    }
    return vapJsonContent;
}


VapFileInfo* _getFileInfoOfVap(char *filePath, bool forceVap) {
    if (!string_end_with(filePath, ".mp4")){
        return NULL;
    }
    struct stat fileStat;
    if(stat(filePath, &fileStat) < 0) {
        return NULL;
    }
    
    cJSON *vapInfo = getVapcInfo(filePath);
    if (vapInfo == NULL && forceVap) {
        return NULL;
    }
    
    VapFileInfo *info = malloc(sizeof(VapFileInfo));
    info->size = fileStat.st_size;
    info->vap_info = vapInfo;
    info->video_info = getMp4Info(filePath);
    info->path = filePath;
    return info;
}
VapFileInfo* getFileInfoOfVap(char *filePath) {
    return _getFileInfoOfVap(filePath, true);
}

VapFileInfo* getVapFileInfoAllowNotVap(char *filePath) {
    return _getFileInfoOfVap(filePath, false);
}

cJSON * getVapFileInfoJson(char * filePath) {
    VapFileInfo *info = getFileInfoOfVap(filePath);
    if (info == NULL) {
        return NULL;
    }
    cJSON *json = cJSON_CreateObject();
    cJSON *videoInfoJson = videoInfnToJson(info->video_info);
    cJSON_AddStringToObject(json, "path", info->path);
    cJSON_AddNumberToObject(json, "size", info->size);
    cJSON_AddItemToObject(json, "vap_info", info->vap_info);
    cJSON_AddItemToObject(json, "video_info", videoInfoJson);
    free(info);
    free(info->vap_info);
    return json;
}

void writeUInt32BE(uint8_t *buffer, uint32_t value) {
    buffer[0] = (value >> 24) & 0xFF;
    buffer[1] = (value >> 16) & 0xFF;
    buffer[2] = (value >> 8) & 0xFF;
    buffer[3] = value & 0xFF;
}

int addVapcToMp4File(char * outPath, char *vapc) {
    BoxArray allBoxes = getVapBoxes(outPath);
    for (int i = 0; i < allBoxes.length; i++) {
        if (strcmp(allBoxes.boxes[i]->boxType, "vapc") == 0) {
            return -1;
        }
    }

    uint64_t vapcLen = strlen(vapc) + 1;
    uint64_t vapcBoxSize = 8 + vapcLen;
    uint8_t *vapcBoxBuffer = malloc(vapcBoxSize);
    writeUInt32BE(vapcBoxBuffer, (uint32_t)vapcBoxSize);

    memcpy(vapcBoxBuffer + 4, "vapc", 4);
    memcpy(vapcBoxBuffer + 8, vapc, vapcLen);

    char tempFilePath[256];
    sprintf(tempFilePath, "%s.temp", outPath);
    FILE *fd = fopen(tempFilePath, "w");
    FILE *oldFd = fopen(outPath, "r");
    uint64_t position = 0;
    for (int i = 0; i < allBoxes.length; i++) {
        Mp4Box *oldBox = allBoxes.boxes[i];
        char *oldBoxBuffer = readCharFromFile(oldFd, oldBox->start, oldBox->size);
        fwrite(oldBoxBuffer, sizeof(char), oldBox->size, fd);
        position = oldBox->start + oldBox->size;
    }
    fwrite(vapcBoxBuffer, sizeof(char), vapcBoxSize, fd);
    fclose(fd);
    fclose(oldFd);
    
    BoxArray newAllBox = getVapBoxes(tempFilePath);
    int boxOk = 0;
    for (int i = 0; i < newAllBox.length; i++) {
        if (strcmp(newAllBox.boxes[i]->boxType, "vapc") == 0) {
            boxOk = 1;
            break;
        }
    }
  
    if (boxOk) {
        remove(outPath);
        rename(tempFilePath, outPath);
        return 0;
    } else {
        remove(tempFilePath);
        return -2;
    }
}
